src/main/java/org/jboss/set/pull/processor/Main.java

```diff
@@ -0,0 +1,176 @@
+        logger.info("initializing....");
+        try (Aphrodite aphrodite = Aphrodite.instance();
+                ClosableHackForExecutor executor = new ClosableHackForExecutor(Executors.newFixedThreadPool(12));) {
+            if (parsedStreams.isEmpty()) {
+                // this shouldnt happen
+                logger.info("No streams specified, can't work like that, make up your mind");
+                return;
+            } else {
+                // XXX: this is a bit scechy, but should do for first iteration
+                matchStreams(aphrodite,parsedStreams);
+            }
+
+            if (writePermittedStreams != null && !writePermittedStreams.isEmpty()) {
+                matchStreams(aphrodite,writePermittedStreams);
+            }
+
+            logger.info("loading evaluators:");
+            ServiceLoader<Evaluator> evaluatorsServiceLoader = ServiceLoader.load(Evaluator.class);
+            final List<Evaluator> evaluatorServices = new ArrayList<Evaluator>();
+
+            for (Evaluator evaluator : evaluatorsServiceLoader) {
+                evaluatorServices.add(evaluator);
+                logger.info("...." + evaluator.getClass().getSimpleName());
+            }
+
+            logger.info("loading actions:");
+            ServiceLoader<Action> actionServiceLoader = ServiceLoader.load(Action.class);
+            final List<Action> actionServices = new ArrayList<Action>();
+            for (Action action : actionServiceLoader) {
+                actionServices.add(action);
+                logger.info("...." + action.getClass().getSimpleName());
+            }
+
+            ServiceLoader<Processor> processors = ServiceLoader.load(Processor.class);
+            logger.info("configuring processors:");
+            // yeah, two loops, could be done in one go, but KISS is life.
+            for (Processor processor : processors) {
+                logger.info("...." + processor.getClass().getName());
+                final ProcessorPhase processorPhase = processor.getPhase();
+                // find actions and evaluators that can be run by processor
+                final List<Action> filteredActions = actionServices.stream().filter(a -> a.support(processorPhase))
+                        .collect(Collectors.toList());
+                final List<Evaluator> filteredEvaluators = evaluatorServices.stream().filter(a -> a.support(processorPhase))
+                        .collect(Collectors.toList());
+                final ProcessorConfig processorConfig = new ProcessorConfig(filteredEvaluators, filteredActions, parsedStreams, writePermittedStreams,
+                        aphrodite, executor.executorService, rootDir, performWriteOperations);
+                processor.init(processorConfig);
+            }
+
+            logger.info("executing processors: ");
+            for (Processor processor : processors) {
+                logger.info("...." + processor.getClass().getName());
+                processor.process();
```

Comment:
// yeah, two loops, could be done in one go, but KISS is life.
So we have clear phase of setting up everything and clear phase of doing something.